add support for Post method in SUGAR

hchoura opened the pull request.

KateMengjunXia committed:
Initial NotificationSetting

KateMengjunXia committed:
Adding setGuid to contificationConfigDAO

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

Commit:
add support for Post method in SUGAR

Commit:
add support for Post method in SUGAR

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -6,29 +6,37 @@
+
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
```
Why are we using this library?

nanoscott replied:
@kgrgreer .. how do we parse JSON into a map of key value pairs?

hchoura replied:
I switch it to foam MapParser

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -299,9 +321,40 @@ protected Object setParameterValue(X x, HttpServletResponse resp, PrintWriter ou
+        case "foam.core.X":
+          arg = x;
+          break;
+        default:
```
The whole reason for doing this is so that more complex objects can be passed into the SUGAR API. Can you figure out how to parse FObjects here. An example of a call that you should be able to parse is the AgentAuthService.canActAs where the second and third parameters are User objects and not simple string/int/bools .. you will need to figure out how to parse these here as well.

hchoura replied:
We will create object based on the method type.

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +136,28 @@ public void execute(X x) {
               paramTypes[j] = pArray[j].getType();
-              arglist[j] = p.getParameter(pArray[j].getName());
+
+              if ( p.getParameter(pArray[j].getName()) != null ) {
```
I think we can simply remove the ability for users to pass in method parameters in the GET URL. This code should no longer be necessary. So it would be required that whenever a method has parameters, the caller uses a POST call with a message body in order to pass in the parameters.

hchoura replied:
I prefer to have the post and the get method for SUGAR.

nanoscott replied:
GET still works as long as there are no method parameters in the URL. We can remove this code and force users to pass in method parameters through a POST body.

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +136,28 @@ public void execute(X x) {
 
               if ( typeName.equals("foam.core.X") ) {
                 arglist[j] = x;
               } else if ( ! SafetyUtil.isEmpty(p.getParameter(pArray[j].getName())) ) {
```
This should be able to be remove too.

Nauna committed:
file doesn't exist so removing

kgrgreer committed:
Merge pull request #2777 from Nauna/removeWrongFile

file doesn't exist so removing

KateMengjunXia committed:
Implemented sendNotification

KateMengjunXia committed:
Implemented sendNotification

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

Commit:
using foam parser.
create object base on the method type.

Commit:
using foam parser.
create object base on the method type.

Commit:
organize import.

Commit:
organize import.

Commit:
add missing header

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -45,6 +56,12 @@ public void execute(X x) {
     String              methodName     = p.getParameter("method");
     String              interfaceName  = p.getParameter("interfaceName");
 
+    String data = p.getParameter("data");
```
what happens when data is not provided? i.e. data is null?

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +141,59 @@ public void execute(X x) {
+              } else if ( paramName != null ) {
+                //post method
+                //TODO array
+                if ( typeName.equals("double") || typeName.equals("boolean") || typeName.equals("int") || typeName.equals("long") || typeName.equals("long") || typeName.equals("java.lang.String") ) {
```
You shouldn't do it this way. You can add all of the supported types to a Set (which could be made a constant), then do a Set.contains here instead of having this long if statement.

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +141,59 @@ public void execute(X x) {
+                //post method
+                //TODO array
+                if ( typeName.equals("double") || typeName.equals("boolean") || typeName.equals("int") || typeName.equals("long") || typeName.equals("long") || typeName.equals("java.lang.String") ) {
+                  arglist[j] = mapPostParam.get(paramName);
```
won't you still have the casting problem here where all numbers are parsed as Doubles? If not, how is the MapParser able to distinguish between ints and longs and doubles?

Commit:
move predicate to foam

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +141,59 @@ public void execute(X x) {
+                    // For Enum Parameter
+                    if ( objClass != null && objClass.isEnum() ) { 
+                      for ( int t = 0 ; t < objClass.getEnumConstants().length ; t++ ) {
+                        if ( objClass.getEnumConstants()[t].toString().equals(mapPostParam.get(paramName)) ) {
```
what if the enum value is passed in as an integer?

nanoscott replied:
use SafetyUtil.equals here

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +141,59 @@ public void execute(X x) {
+                        }
+                      }
+                    } else {
+                      FObject             obj = (FObject) objClass.newInstance();
```
can't objClass be null here?

nanoscott commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -123,19 +141,59 @@ public void execute(X x) {
+                      Map<String, String> mParam   = (Map) mapPostParam.get(paramName);
+
+                      for ( Map.Entry<String, String> entry : mParam.entrySet() ) {
+                        obj.setProperty(entry.getKey(), entry.getValue());
```
so this works even if the property is not a string?

nanoscott replied:
what about nested objects?

Commit:
move predicates and actions to foam

Commit:
improve documentation with generic examples

nanoMichal committed:
Update SendNotification.js

change E to e.

Commit:
add improved documentation

Commit:
remove else

Commit:
remove unneccessary elses

Commit:
Merge branch 'morePredicates' of https://github.com/nanoMichal/foam2 into morePredicates

kgrgreer committed:
Merge pull request #2778 from nanoMichal/predicates

Predicates

nanoMichal committed:
Merge branch 'master' into morePredicates

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

jlhughes committed:
Merge pull request #2779 from nanoMichal/morePredicates

move predicates and actions to foam

Commit:
move predicates

jlhughes committed:
Merge pull request #2782 from nanoMichal/evenmorePredicatesToMove

move predicates

Commit:
Prevent hidden edit button

kgrgreer committed:
Merge pull request #2783 from KernelDeimos/CPF-3377/PreventHiddenEditButton

[CPF-3377] Prevent hidden edit button

michaelconnor468 committed:
EasyDAO made Easy

michaelconnor468 committed:
addressed stuff

Commit:
sectioned prop mode to slot

kgrgreer committed:
Merge pull request #2785 from zahnno/slot-mode-on-section-prop

Sectioned Detail Property Mode

jlhughes committed:
Merge pull request #2784 from michaelconnor468/EasyDecoratorChain

EasyDAO made Easy

jlhughes committed:
Explicitly fclone find 'new' object.

CPF-3384 #time 1h

jlhughes committed:
fclone find object before copyFrom

CPF-3384 #time 1h

KateMengjunXia committed:
Implemented notify method

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

Commit:
Don't include trailing commas in CSV output

Also includes general code style fixes.

Commit:
Formatting

* Invert condition and return early in simplifyPredicate instead of
putting all of the logic inside an if block.
* Clarify a few comments
* Simplify code in planSelect

jlhughes committed:
Merge pull request #2786 from foam-framework/CPF-3384/FcloneFixes-DAOSkeleton

fclone find object before copyFrom

jlhughes committed:
Merge pull request #2787 from foam-framework/CPF-3384/FcloneFixes-Ruler

Explicitly fclone find 'new' object.

xuerongNanopay committed:
add enabledaware to nspec

xuerongNanopay committed:
remove view

jlhughes committed:
Merge pull request #2790 from xuerongNanopay/EnableNSpec

Add enabledaware to nspec

Commit:
Let user select class to create in DAO controller

This commit changes the default detailView for the DAO Controller to
FObjectView instead of DetailView. This allows the user to select the
class that they want to create an instance of.

Since FObjectView's class dropdown is populated by the Strategizer, one
can add StrategyReferences to populate the dropdown. If there are no
alternative strategies for the model being created, then the field to
select the class will be hidden and the base class will be used, (the
'of' of the DAO being controlled), which is consistent with the current
behaviour.

kgrgreer committed:
Merge pull request #2791 from foam-framework/select-subclass-in-create-controller

Let user select class to create in DAO controller

Commit:
Add back commented out code and add TODO

jlhughes committed:
Add Thread.currentThread id to Log messages.

jlhughes committed:
Change from Thread.Id to Thread.name

jlhughes committed:
Merge pull request #2792 from foam-framework/LogCurrentThread

Add Thread.currentThread id to Log messages.

kgrgreer committed:
Merge pull request #2789 from foam-framework/csv-formatting

Various Formatting Improvements

Commit:
Fix bug where objects couldn't be updated via DAO Controller

High Level
==========

I had only tested create yesterday, not viewing or updating, and
because of that I didn't realize that my changes had broken viewing and
editing via the GUI.

This commit fixes the issue.

Technical
=========

Now we only use FObjectView in the create controller, not the update
controller. This allows users to select the subclass when creating a new
instance, but not when editing an existing object.

To respect the DAO Controller's feature of a custom detail view, this
commit updates FObjectView so that it supports changing the view used
for the FObject. In DAOCreateControllerView we set that view to the
custom detail view for the DAO controller if it's available.

kgrgreer committed:
Merge pull request #2793 from foam-framework/fix-dao-update-controller-errors

Fix bug where objects couldn't be updated via DAO Controller

Commit:
Use the MapParser to parse Sugar body in Post method request.

Fix FObjectParser.

Commit:
Use the MapParser to parse Sugar body in Post method request.

Fix FObjectParser.

yij793 committed:
when mode is RO or DISABLED, the label should not clickable

nanoscott commented:
Talked to @hchoura offline and discussed adding two things to this PR:
- Add a test to exercise the SUGAR API
- Update the SUGAR UI to allow for users to paste JSON into the Data field and send the POST properly

yij793 committed:
fix the indentation, change condition using self.getAttribute()

kgrgreer committed:
Merge pull request #2794 from yij793/CPF-3393-Update-checkbox-when-mode-is-RO-or-DISABLED-labe-should-not-clickable

when mode is RO or DISABLED, the label should not clickable

Commit:
Put TimeHSM inputs inline to save space

Commit:
Fix bug where default value of ChoiceView wasn't being respected

Commit:
Set default choice in FObjectView to first choice

Commit:
Use FObjectView as default write view for FObjectProperties

Commit:
Add missing detach listener

Commit:
Fallback on 'of' when no strategies available; Sort class dropdown
in FObjectView alphabetically

Commit:
Add strategy references for cron schedules

kgrgreer committed:
Merge pull request #2797 from foam-framework/cron-ui-and-FObjectView-improvements

Cron UI and FObjectView improvements

kgrgreer committed:
Initial check-in of AssemblinLine.

jlhughes committed:
Do not send email already in FAILED state.

jlhughes committed:
Add t to error log to generate stack trace in log files to help troubleshoot session server box issues.

Nauna committed:
adding a background colour to be used with our logos

jlhughes committed:
Merge pull request #2800 from foam-framework/CPF-3437/SessionServerBoxStackTrace

Add t to error log to generate stack trace in log files to help troub…

jlhughes committed:
Merge pull request #2799 from foam-framework/CPF-3434/DoNotSendFailedEmail

Do not send email already in FAILED state.

kgrgreer committed:
Initial check-in of an asynchronous AssemblyLine.

kgrgreer committed:
Add missing package.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

kgrgreer committed:
Fix service name typo.

Nauna committed:
name change on property

KateMengjunXia committed:
Added relationships between user and notification setting

KateMengjunXia committed:
Merge remote-tracking branch 'upstream/master' into NotificationSetting

jlhughes committed:
Merge pull request #2765 from KateMengjunXia/NotificationSetting

Notification setting models

jlhughes committed:
Add permissions for strategyDAO and strategyreference.read

jlhughes committed:
Revert "Notification setting models"

jlhughes committed:
Merge pull request #2804 from foam-framework/revert-2765-NotificationSetting

Revert "Notification setting models"

Nauna committed:
Update src/foam/nanos/theme/Theme.js

Co-Authored-By: Troy Wolters <troy@nanopay.net>

jlhughes committed:
Remove jrl endings to they work in foam

jlhughes committed:
Missed adding the update to src/servicees

kgrgreer committed:
Initial check-in of fast timestamp creator.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

kgrgreer committed:
Merge pull request #2801 from Nauna/addPropertiesForLogo

adding a background colour to be used with our logos

jlhughes committed:
Add permission for strategizer service

kgrgreer committed:
Save context in AsyncAssemblyLine.

kgrgreer committed:
Formatting and TODO.

kgrgreer committed:
Formatting.

kgrgreer committed:
Replace foam.util.Dates with two classses for creating fast timestamps.

kgrgreer committed:
Fix loggers not outputting milliseconds portion of timestamps correctly.

Commit:
Add UI for Post method in SUGAR

Commit:
Add UI for Post method in SUGAR

Commit:
format

Commit:
format

kgrgreer committed:
Initial check-in.

kgrgreer committed:
Initial check-in.

nanoscott commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -279,12 +242,12 @@ foam.CLASS({
+        if ( this.sugarURL !== '') {
           var req = this.HTTPRequest.create({
-            url: window.location.protocol + '//' + window.location.hostname + ':' + window.location.port + this.postURL + this.postData+ "&sessionId=" + localStorage.defaultSession,
+            url: window.location.protocol + '//' + window.location.hostname + ':' + window.location.port + this.sugarURL + '?data=' + this.postData+ '&sessionId=' + localStorage.defaultSession,
```
postData should not be included twice in this .. if it is included in the URL ensure that it is encoding properly so that browsers can handle the request.

nanoscott reviewed (approved):
Just one minor comment. Please address ... otherwise it looks good.

kgrgreer commented on src/foam/lib/json/FObjectParser.java:
```diff
@@ -46,7 +46,13 @@ public PStream parse(PStream ps, ParserContext x) {
               subParser = EnumParserFactory.getInstance(c);
             } else {
-              Object obj = ((X) x.get("X")).create(c);
+              Object obj = null;
```
Why do we do this rather than just always provide the X when using FObjectParser?

hchoura replied:
Can you explain more?

nanoscott replied:
Please follow-up with Kevin directly to address this comment.

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -208,7 +208,7 @@ foam.CLASS({
       name: 'sugarURL',
       label: 'URL',
-      hidden: false,
+      hidden: true,
```
Why do we hide this?

hchoura replied:
Since we adopt the post method for all SUGAR service. The URL will be not useful. Instead, the bottom will send the request.

nanoscott replied:
looks like you are using the URL below to perform the POST, no?

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -217,54 +217,17 @@ foam.CLASS({
-        }
-
-        return encodeURI(url + this.postData);
+        this.postData=this.data.toString();
```
Formatting:
this.postData = this.data.toString();

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -279,12 +242,12 @@ foam.CLASS({
       label: 'Send POST Request',
       code: async function() {
-        if ( this.postURL !== '' ) {
+        if ( this.sugarURL !== '') {
```
Why was the space removed?

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -40,20 +50,33 @@ public void execute(X x) {
+    String              data           = p.getParameter("data");
 
     try {
+      if ( SafetyUtil.isEmpty(data) ) {
```
Is this an error if you call a method with no arguments?

hchoura replied:
data contain the service, method, interfaceName and set of parameters. It should not be empty.

Commit:
Formatting

Commit:
Formatting

lchanmann committed:
Fix property reference parser white spacing

jlhughes committed:
Add create, update, remove permissions for completeness.

jlhughes committed:
Merge pull request #2803 from foam-framework/strategypermissions

Add permissions for strategyDAO and strategyreference.read

Commit:
Fix GUI for editing appConfig NSpec

kgrgreer committed:
Added U2 demo on creating a multi-part text field.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

kgrgreer committed:
Update syntax.

jlhughes committed:
Merge pull request #2811 from lchanmann/FixPropertyReferenceParser

Fix property reference parser white spacing

kgrgreer committed:
Merge pull request #2813 from foam-framework/fix-appConfig-editing

Fix GUI for editing appConfig NSpec

Commit:
add sugar service test

Commit:
add sugar service test

nanoscott commented on src/sugars:
```diff
@@ -0,0 +1 @@
+p({"class":"foam.nanos.dig.SUGAR","id":"checkUserTest","data":"{\n    \"service\": \"auth\",\n    \"interfaceName\": \"foam.nanos.auth.AuthService\",\n    \"method\": \"checkUser\",\n    \"X\": \"X\",\n\n    \"user\": {\n    \t\"class\":\"foam.nanos.auth.User\",\"id\":1348,\"enabled\":true,\"loginEnabled\":true,\"firstName\":\"admin\",\"middleName\":\"\",\"lastName\":\"admin\",\"organization\":\" ...
```
Please use the new multi-line format here so this is readable.

yij793 committed:
add error style for RichChoiceView-selection-view, remove selection-view border since it will overwirte the error style

kgrgreer committed:
Initial check-in.

kgrgreer committed:
Initial check-in.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

jlhughes committed:
Merge pull request #2817 from yij793/CPF-2255/foam-RichChoice-section-view-error-is-not-be-added

add error style for RichChoiceView-selection-view, remove selection-v…

Commit:
add multi-line format to the journal file.

Commit:
add multi-line format to the journal file.

Commit:
add test in the journal test

Commit:
add test in the journal test

Commit:
move ApiTestBase to foam

Commit:
Merge branch 'PostMethodSupport' of https://github.com/foam-framework/foam2.git into PostMethodSupport
